book/src/solana-scripts.md

```diff
@@ -0,0 +1,253 @@
+# Solana Scripts
+
+In Solana programs have absolute control over the state assigned to them.  The
+goal of this architecture is provide a simple way for on-chain programs to
+interact with other programs without violating the state ownership guarantees.
+
+Contract composition has led to multiple multi-million dollar bugs in Ethereum
+
+* https://www.parity.io/security-alert-2/
+
+A few solutions that have been tried:
+
+* Eos: message queues
+* Ethereum: implementation delegation
+* Libra: resources
+
+The ideal solution to this problem is an infallible Trusted Third Party. For
+example, clients that want to bet on an outcome of an on-chain game of
+Tic-Tac-Toe, transfer their tokens to the TTP client, and the TTP transfers the
+prize to the winner at the end of the game.
+
+Scripts are on-chain programs that act as an infallible TTP client.
+
+## Scripts
+
+To implement an infallible TTP, scripts behave as on-chain clients, with an
+on-chain wallet that has access to script unique keypairs that only the script
+can sign with.  Additionally, like regular clients, scripts cannot directly
+mutate state or transfer funds, and must do those things via calls to
+instructions.
+
+Scripts are executable bytecode that run in an environment with additional
+features and restrictions than programs.
+
+Scripts have the following additional features:
+* `process_instruction`, a synchronous call to process an instructions.
+* `keypair_pubkey`, generate a persistent key that only this script can sign with.
+* `sign_instruction`, sign an instruction with a script key.
+
+Since scripts are clients, they cannot:
+* Directly mutate state, or directly move funds between accounts.
+
+Scripts rely on programs to move funds or mutate state, and act purely as
+clients.
+
+## Secure Script Execution
```

Comment:
How is this different from a non-script case where two clients sign transactions based on the current state and one gets executed before the other?  Won't the later transaction may be executed on state that has evolved since that transaction was signed.

Reply:
@jackcmay with a single program, the program controls all the state transitions.  with meta program, F calls G, H, users can independently call G, and H.  So F doesn't control all the state transitions.

Reply:
II think this is related to my question below.  Is the whole script executed atomically in terms of chain state?

Reply:
@jackcmay yep, in the context of a single transaction.

Reply:
@aeyakovenko Wouldn't your scenario result in an `AccountInUse` error?  Or are you thinking that F and the independent calls to G and H are all in the same TX?

Reply:
All the calls to external programs made by the script are in the clients transaction instruction vector, in the order the script will make them.